datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -644,6 +649,20 @@ private void handleAssignmentChange(boolean isDatastreamUpdate) throws TimeoutEx
         .flatMap(Collection::stream)
         .collect(Collectors.toMap(DatastreamTask::getDatastreamTaskName, Function.identity())));
 
+    if ((totalTasks - submittedTasks) > 0) {
+      _log.warn("Failed to submit {} tasks from currentAssignment. Queueing onAssignmentChange event again", totalTasks - submittedTasks);
+      // Update the metric and queue the event only once for all the tasks that failed to be initialized
+      if (isDatastreamUpdate) {
+        _metrics.updateKeyedMeter(CoordinatorMetrics.getKeyedMeter(
+            CoordinatorEvent.EventType.HANDLE_DATASTREAM_CHANGE_WITH_UPDATE), 1);
+        _eventQueue.put(CoordinatorEvent.createHandleDatastreamChangeEvent());
+      } else {
+        _metrics.updateKeyedMeter(CoordinatorMetrics.getKeyedMeter(
+            CoordinatorEvent.EventType.HANDLE_ASSIGNMENT_CHANGE), 1);
+        _eventQueue.put(CoordinatorEvent.createHandleAssignmentChangeEvent());
+      }
+    }
+
```

Comment:
I refactored the code a bit from my previous patch.
As we had discussed to update the metric count only once irrespective of the number of task failures I thought I could do that and also queue the corresponding event together.
To check if there were any task failures I am checking for a diff in task count instead of previous _failedDatastreamTasks set/map that I was using earlier. I thought that it would avoid having to maintain a consistent state for this additional variable, however this way I end up (potentially) modifying "currentAssignment" in the dispatch method. However I am not sure if this is against any coding guidelines that we follow. Pls let me know what is the correct way to handle this. Or maybe I can simply rename "dispatchAssignmentChangeIfNeeded" to "dispatchAndTrimAssignmentChangeIfNeeded" ?